semantic-analysis/src/Analysis/Syntax.hs

```diff
@@ -0,0 +1,180 @@
+import qualified Data.ByteString.Lazy as B
+import           Data.Function (fix)
+import qualified Data.IntMap as IntMap
+import           Data.Monoid (First(..))
+import           Data.Text (Text, pack, unpack)
+import qualified Data.Vector as V
+
+class Syntax rep where
+  iff :: rep -> rep -> rep -> rep
+  noop :: rep
+
+  bool :: Bool -> rep
+  string :: Text -> rep
+
+  throw :: rep -> rep
+
+  let_ :: Name -> rep -> (rep -> rep) -> rep
+
+
+-- Pretty-printing
+
+newtype Print = Print { print_ :: ShowS }
+
+instance Show Print where
+  showsPrec _ = print_
+
+instance Semigroup Print where
+  Print a <> Print b = Print (a . b)
+
+instance Monoid Print where
+  mempty = Print id
+
+instance Syntax Print where
+  iff c t e = parens (str "iff" <+> c <+> str "then" <+> t <+> str "else" <+> e)
+  noop = parens (str "noop")
+
+  bool b = parens (str (if b then "true" else "false"))
+  string = parens . text
+
+  throw e = parens (str "throw" <+> e)
+
+  let_ n v b = parens (str "let" <+> name n <+> char '=' <+> v <+> str "in" <+> b (name n))
+
+str :: String -> Print
+str = Print . showString
+
+text :: Text -> Print
+text = str . unpack
+
+char :: Char -> Print
+char = Print . showChar
+
+parens :: Print -> Print
+parens p = char '(' <> p <> char ')'
+
+(<+>) :: Print -> Print -> Print
+l <+> r = l <> char ' ' <> r
+
+infixr 6 <+>
+
+name :: Name -> Print
+name = text . formatName
+
+
+-- Abstract interpretation
+
+eval0 :: Interpret m i -> m i
+eval0 = fix eval
+
+eval :: (Interpret m i -> m i) -> (Interpret m i -> m i)
+eval eval (Interpret f) = f eval
+
+newtype Interpret m i = Interpret { interpret :: (Interpret m i -> m i) -> m i }
+
+instance (Has (Env addr) sig m, HasLabelled Store (Store addr val) sig m, Has (Dom val) sig m) => Syntax (Interpret m val) where
+  iff c t e = Interpret (\ eval -> do
+    c' <- eval c
+    dif c' (eval t) (eval e))
+  noop = Interpret (const dunit)
+
+  bool b = Interpret (\ _ -> dbool b)
+  string s = Interpret (\ _ -> dstring s)
+
+  throw e = Interpret (\ eval -> eval e >>= ddie)
+
+  let_ n v b = Interpret (\ eval -> do
+    v' <- eval v
+    let' n v' (eval (b (Interpret (pure (pure v'))))))
+
+
+-- Macro-expressible syntax
+
+let' :: (Has (Env addr) sig m, HasLabelled Store (Store addr val) sig m) => Name -> val -> m a -> m a
+let' n v m = do
+  addr <- alloc n
+  addr .= v
+  bind n addr m
+
+
+-- Parsing
+
+parseFile :: Syntax rep => FilePath -> IO (Either (A.JSONPath, String) (Maybe rep))
+parseFile path = do
+  contents <- B.readFile path
+  pure $ snd <$> A.eitherDecodeWith A.json' (A.iparse parseGraph) contents
+
+parseGraph :: Syntax rep => A.Value -> A.Parser (IntMap.IntMap rep, Maybe rep)
+parseGraph = A.withArray "nodes" $ \ nodes -> do
+  (untied, First root) <- foldMap (\ (k, v, r) -> ([(k, v)], First r)) <$> traverse (A.withObject "node" parseNode) (V.toList nodes)
+  let tied = fix (\ tied -> ($ tied) <$> IntMap.fromList untied)
+  pure (tied, ($ tied) <$> root)
```

Comment:
Here's where we "tie the knot," taking the non-recursive higher-order data and applying it recursively to the result of applying it recursively to the result of… you get where I'm going with this.

We also need to do the same for the root node, which is also defined over the completed graph.

Reply:
This is lovely; can you move this explanation into a code comment?